src/lib/parser.c

```diff
@@ -701,39 +704,45 @@ static RdbStatus allocFromCache(RdbParser *p,
 }
 
 static inline RdbStatus unpackList(RdbParser *p, unsigned char *lp) {
-    char dummy, tmp = 'x', *itemStart, *itemEnd;
+    char dummy, tmp = 'x', *item, *itemEnd;
     unsigned char *eptr;
     unsigned int vlen;
     long long vll;
 
     eptr = lpFirst( lp);
     while (eptr) {
-        itemStart = (char *)lpGetValue(eptr, &vlen, &vll);
+        item = (char *) lpGetValue(eptr, &vlen, &vll);
         BulkInfo *binfo;
 
-        if (itemStart) {
+        if (item) {
             /* The callback function expects a native string that is terminated
              * with '\0'. However, the string we have is packed without
              * termination. To avoid allocating a new string, we can follow these
              * steps:
-             * 1. Save the last character that comes at the end of the referenced
-             *    string in a temporary char (tmp).
-             * 2. allocFromCache(RQ_ALLOC_APP_BULK_REF) will:
+             * 1. Save the last character that comes after the end of the packed
+             *    string in a temporary char (tmp). This is valid and not beyond the
+             *    size of the allocation since in listpack it is guaranteed to have
+             *    a terminating byte at the end, which makes it safe.
+             *
+             * 2. Then allocate from cache `RQ_ALLOC_APP_BULK_REF` which will:
              *    - Set last character '\0' to terminate the string.
-             *    - Mark the string as a referenced bulk allocation (placement-new alloc)
-             *      Note: IF app expects APP_BULK, then it is not possible to return
-             *            a reference. A new memory will be allocated instead and gets a copy
-             * 3. invoke CALL_HANDLERS_CB that will:
-             *    - Supply the bulk to callbacks
-             *    - Finalize by restoring original char (from tmp) */
-            tmp = *(itemEnd = itemStart + vlen);
-
-            IF_NOT_OK_RETURN(allocFromCache(p, vlen, RQ_ALLOC_APP_BULK_REF, itemStart, &binfo));
+             *    - And mark the string as a referenced bulk allocation
+             *      (Note, iF app expects APP_BULK, then it is not possible to return
+             *      a reference and a new memory will be allocated instead with proper
+             *      termination of '\0').
+             *
+             * 3. And lastly invoke CALL_HANDLERS_CB that will:
+             *    - Supply the referenced, or copied, application bulk to callback
+             *    - And finalize by restoring original char from `tmp`.
+             */
+            itemEnd = item + vlen;
+            tmp = *itemEnd;
```

Comment:
i think the other de-reference of itemEnd should also move to a line of its own.
that is unless i misunderstand (didn't bother to check) how that CALL_HANDLERS_CB works.
```
                         *itemEnd = tmp,   /* <<< finalize: restore modified char */
```

Reply:
`*itemEnd = tmp` is given as `finalize` operation to `CALL_HANDLERS_CB()` which will be called either after successful execution or on `return` failure. So this is the only way to gurantee to revert temporary write either way. 

Having said that, following an error from callback there shouldn't be any further attempt to use the `item` and we can avoid revert. Though (it might be unclear in case of debugging).

i don't have strong opinion either way. WDYT?

Reply:
i think we should revert. even if just to make a point.
i thought that part of the statement is an argument to a function that gets evaluated.
i guess it can't be piece of code executed by a callback, which means it's just a block inside an `if`?
sorry, too lazy and busy to clone the code and take a deeper look